### ISSUE TITLE
HMC fails to 'draw' from standard gaussian target

### ISSUE THREAD

**Issue by wiep** [blackjax-devs/blackjax#46]
blackjax currently fails to target the standard normal distribution with HMC. in my example, the empirical mean is close to 0  but the empirical variance is close to 0.5. similar values arise when employing nuts with and without stan warmup.

running
```py
import numpy as np
import jax
import jax.numpy as jnp
import blackjax.hmc as hmc
import matplotlib.pyplot as plt

potential = lambda x: -jax.scipy.stats.norm.logpdf(x, loc=0.0, scale=1.0).squeeze()
initial_position = np.array([1.0,])
initial_state = hmc.new_state(initial_position, potential)
initial_state

inv_mass_matrix = 0.1 * jnp.ones_like(initial_position)
num_integration_steps=100
step_size=1e-2

hmc_kernel = hmc.kernel(
    potential,
    step_size=step_size,
    inverse_mass_matrix=inv_mass_matrix,
    num_integration_steps=num_integration_steps
)

hmc_kernel = jax.jit(hmc_kernel)

def inference_loop(rng_key, kernel, initial_state, num_samples):
    def one_step(state, rng_key):
        state, _ = kernel(rng_key, state)
        return state, state

    keys = jax.random.split(rng_key, num_samples)
    _, states = jax.lax.scan(one_step, initial_state, keys)

    return states

rng_key = jax.random.PRNGKey(0)
states = inference_loop(rng_key, hmc_kernel, initial_state, 50_000)

samples = states.position.block_until_ready()
print(np.mean(samples, axis=0))
print(np.var(samples, axis=0))
plt.plot(samples)
plt.show()
```
gives
```
[-0.00325776]
[0.47831735]
```

it seems the behaviour changed with 7be282232d87d93a15d5c5ff07648e2eb6f144c6. in particular, with the change in `blackjax/inference/proposal.py`.
using `p_accept = jnp.clip(jnp.exp(proposal.weight), a_max=1)` gives with the above code
```
[-0.00980166]
[0.9992622]
```

similar, with the parent commit (i.e., c6f75e91217f3336c6b0cea33d53efa5c8c843b5)
```py
import numpy as np
import jax
import jax.numpy as jnp
import blackjax.hmc as hmc
import matplotlib.pyplot as plt

potential = lambda x: -jax.scipy.stats.norm.logpdf(x, loc=0.0, scale=1.0).squeeze()
initial_position = np.array([1.0,])
initial_state = hmc.new_state(initial_position, potential)
initial_state

inv_mass_matrix = 0.1 * jnp.ones_like(initial_position)
num_integration_steps=100
step_size=1e-2
params = hmc.HMCParameters(
    step_size=step_size,
    inv_mass_matrix=inv_mass_matrix,
    num_integration_steps=num_integration_steps
)
hmc_kernel = hmc.kernel(potential, params)

def inference_loop(rng_key, kernel, initial_state, num_samples):
    def one_step(state, rng_key):
        state, _ = kernel(rng_key, state)
        return state, state

    keys = jax.random.split(rng_key, num_samples)
    _, states = jax.lax.scan(one_step, initial_state, keys)

    return states

rng_key = jax.random.PRNGKey(0)
states = inference_loop(rng_key, hmc_kernel, initial_state, 50_000)

samples = states.position.block_until_ready()
print(np.mean(samples, axis=0))
print(np.var(samples, axis=0))
plt.plot(samples)
plt.show()
```
results in
```
[-0.00980166]
[0.9992622]
```

**Comment by rlouf** [comment]
Thank you for the bug report! What bothers me is that the computation of `p_accept` in this function should not affect NUTS.

Could you confirm that you also observe the bug for NUTS at the parent commit?

**Comment by wiep** [comment]
i just tried nuts on the current main branch. i'll try to find time tonight and check if nuts was working on c6f75e91217f3336c6b0cea33d53efa5c8c843b5.

**Comment by junpenglao** [comment]
Thanks for the feedback - yeah we need to add a test with monte carlo central limit theorem for this.

**Comment by wiep** [comment]
i am pretty but not 100% sure that the example code is correct. i also didn't spend much time trying to optimize the parameter.

nuts seems to work on v1 (8209172165b66b7b4c08807b3bef49f1b1a3bac6) in the sense that the sample mean is close to zero and the sample variance is kind of close to unity (1.17 but this might be due to bad parameters for NUTS). in commit cfeffb1ed4ed95fcafa8720ddfe2ea29029c4a3b, i can notice different results compared to v1: the sample variance is now 0.33691323

```py
import numpy as np
import jax
import jax.numpy as jnp
import blackjax.nuts as nuts
import matplotlib.pyplot as plt

potential = lambda x: -jax.scipy.stats.norm.logpdf(x, loc=0.0, scale=1.0).squeeze()
initial_position = np.array([1.0,])
initial_state = nuts.new_state(initial_position, potential)
initial_state

step_size=0.1
params = nuts.NUTSParameters(
    step_size=step_size,
    inv_mass_matrix = 1 * jnp.ones_like(initial_position)
)
nuts_kernel = nuts.kernel(potential, params)
nuts_kernel = jax.jit(nuts_kernel)

def inference_loop(rng_key, kernel, initial_state, num_samples):
    def one_step(state, rng_key):
        state, _ = kernel(rng_key, state)
        return state, state

    keys = jax.random.split(rng_key, num_samples)
    _, states = jax.lax.scan(one_step, initial_state, keys)

    return states

rng_key = jax.random.PRNGKey(0)
states = inference_loop(rng_key, nuts_kernel, initial_state, 50_000)

samples = states.position.block_until_ready()
print(np.mean(samples, axis=0))
print(np.var(samples, axis=0))
plt.plot(samples)
plt.show()
```

| commit | 8209172165b66b7b4c08807b3bef49f1b1a3bac6 | fe6ff69144b99467a9b343b9a331d4b067602e31 | cfeffb1ed4ed95fcafa8720ddfe2ea29029c4a3b |
|--------|------------------------------------------|------------------------------------------|------------------------------------------|
| output | [0.0027857] [1.1750876]                  | [0.0027857] [1.1750876]                  | [0.0067384] [0.33691323]                 |

**Comment by rlouf** [comment]
I'll track this one down, it might be the reason why adaptations fails for the variance of a gaussian target. I will add the extra tests you were talking about @junpenglao with the bug fix.

**Comment by rlouf** [comment]
I have found a bug which *may* explain what you observe (I will try to run the corrected code later tonight).

In the `init` function for the proposal, `position` is passed to the kinetic energy instead of `momentum`. The fact that we can pass both is a remnant of the code on which BlackJAX is based, where I tried to implement the SoftAbs metric. I will also remove that possibility for now.

**Comment by rlouf** [comment]
The first commit on #47 fixes it. I will do a little refactor, add an extra test like @junpenglao suggested and will release a patch.

**Comment by hriebl** [comment]
Thank you for your work, @rlouf!

**Comment by wiep** [comment]
thank you very much for the fix ❤️
